guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+If you are using "and" in the name of a test, this could indicate the test involves too much behavior and may need to be broken up.
+
+🚫
+
+``` typescript
+it("the block tracker is started and returns a promise that resolves to a block number", () => {
+  // ...
+});
+```
+
+✅
+
+``` typescript
+it("starts the block tracker", () => {
+  // ...
+});
+
+it("returns a promise that resolves to a block number", () => {
+  // ...
+});
+```
+
+## Don't directly test private code
+
+Private code is a part of an interface that is not intended to be used by consumers of that interface. This could mean:
+
+- a function or class not exported from a module
+- a method whose name starts with `#` (the syntax for [ECMAScript private fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields))
+- a method whose name starts with `_` (an informal approach popular prior to private fields)
+- a method preceded by the `private` keyword in TypeScript
+- a function or method tagged with the `@private` JSDoc tag
+
+Test code marked as private as though it were directly part of the public interface where it is executed. For instance, although you might write the following:
+
+``` typescript
+// block-tracker.ts
+import { request } from '...';
+
+export class BlockTracker extends EventEmitter {
+  isRunning = false;
+
+  currentBlock: Block | null = null;
+
+  subscriptionId: number | null;
+
+  async stop() {
+    if (!this.isRunning) {
+      return;
+    }
+    this.isRunning = false;
+    this.#startClearCurrentBlockTimer();
+    await this.#end();
+  }
+
+  #startClearCurrentBlockTimer() {
+    setTimeout(() => {
+      this.currentBlock = null;
+    }, 1000);
+  }
+
+  #end() {
+    await request('eth_unsubscribe', this.subscriptionId);
+  }
+}
+```
+
+this is what consumers see:
+
+``` typescript
+// block-tracker.ts
+import { request } from '...';
+
+export class BlockTracker extends EventEmitter {
+  isRunning = false;
+
+  currentBlock: Block | null = null;
+
+  subscriptionId: number | null;
+
+  async stop() {
+    if (!this.isRunning) {
+      return;
+    }
+    this.isRunning = false;
+    setTimeout(() => {
+      this.currentBlock = null;
+    }, 1000);
+    await request('eth_unsubscribe', this.subscriptionId);
+  }
+}
+```
+
+and as a result, this is how the method ought to be tested:
+
+``` typescript
+describe('BlockTracker', () => {
+  describe('stop', () => {
+    it('does not reset the current block if the block tracker is already stopped', () => {
+      // ...
+    });
+
+    it('does not request to unsubscribe if the block tracker is already stopped', () => {
+      // ...
+    });
+
+    it('resets the current block if the block tracker is already stopped', () => {
+      // ...
+    });
+
+    it('requests to unsubscribe if the block tracker is already stopped', () => {
+      // ...
+    });
+  });
+});
+```
+
+## Be conscious of phases
```

Comment:
At face value, this is less of a guideline and more of a sidebar. I wanted to call it out because I reference phases numerous times throughout this document. I do provide a suggestion for highlighting phases below, however, so _that_ is a guideline. I'm wondering whether it would make more sense to reword this guideline to fit the suggestion?

Reply:
Yeah, I agree. I was about to comment on the "All unit tests follow four phases" part, because that isn't really true (some tests repeat setup/exercise/verify steps). But we can say that we _recommend_ writing tests in these phases

Reply:
I've made some changes to this guideline here: [b23c47](https://github.com/MetaMask/contributor-docs/pull/38/commits/7b23c477f506c5ae1f2bb4ca0a96d89cab8d8b0d)

Reply:
Nice!